Scooby-Doo! Night of 100 Frights: Implement New Game

vgm5 opened the pull request:
## What is this fixing or adding?

This adds a wide ranging implementation of Archipelago for the Scooby-Doo Themed Metroid-Vania 3D Platformer, "Scooby-Doo! Night of 100 Frights" for the Nintendo Gamecube

## How was this tested?
This was tested through the publicly available apworld and the feedback received from users who played it

## If this makes graphical changes, please attach screenshots.
No Graphical Changes Made

vgm5 committed:
Initial Commit of apworld

Introduces all necessary files from no100f apworld implementation, minor fixes to become compatible with 0.6.2 and allow for server password entry in client

vgm5 committed:
Polished Various Documents

Updated Setup Guide, Game Info Page, and other WebHost Docs, added Client Icon

vgm5 committed:
Fixed Linux Compatibility Issue

os.startfile will only work from windows, added code to differentiate user system and use different calls if on linux based OS

ScipioWright commented on worlds/no100f/__init__.py:
```diff
@@ -0,0 +1,344 @@
+
+        return item
+
+    def write_spoiler(self, spoiler_handle: TextIO) -> None:
+        return
```
You can just delete this entirely if you're not doing anything with it.
```suggestion
```

vgm5 replied:
Removed next commit

ScipioWright commented on worlds/no100f/docs/setup_en.md:
```diff
@@ -0,0 +1,60 @@
+
+### What is a config file and why do I need one?
+
+YAML Files will tell the generator what your player name is, what game you're playing, and what different settings you want to use.
```
```suggestion
YAML Files will tell the generator what your player name is, what game you're playing, and what different options you want to use.
```

ScipioWright reviewed (commented):
Brief review of common pain points

vgm5 commented on worlds/no100f/Options.py:
```diff
@@ -0,0 +1,118 @@
+    display_name = "Creepy Early"
+    default = 0
+
+class NoLogic(Toggle):
+    """Disables all item Logic | USE AT YOUR OWN RISK"""
+    display_name = "Disable Logic"
+    default = 0
```
I tend to agree, initially this option was added early on to complement the separate standalone randomizer in Industrial Park

At this point in time it just becomes problematic like you describe with hardly any real benefit.

Will remove next commit

vgm5 commented on worlds/no100f/Options.py:
```diff
@@ -0,0 +1,118 @@
+from dataclasses import dataclass
+
+from Options import Toggle, DeathLink, Range, Choice, PerGameCommonOptions
+
+
+class IncludeMonsterTokens(Toggle):
+    """Include Monster Tokens as AP Locations/Items"""
+    display_name = "Include Monster Tokens"
+    default = 1
```
Fixed next commit

vgm5 commented on worlds/no100f/Options.py:
```diff
@@ -0,0 +1,118 @@
+    default = 0
+
+@dataclass
+class NO100FOptions(PerGameCommonOptions):
```
I will add this next commit

vgm5 commented on worlds/no100f/Regions.py:
```diff
@@ -0,0 +1,191 @@
+from BaseClasses import MultiWorld, Region, Entrance
+
+
+def create_region(world: MultiWorld, player: int, name: str, locations=None, exits=None) -> Region:
+    ret = Region(name, player, world)
+    if locations:
+        for location in locations:
+            loc_id = location_table[location]
+            location = NO100FLocation(player, location, loc_id, ret)
+            ret.locations.append(location)
+
+    if exits:
+        for _exit in exits:
+            ret.exits.append(Entrance(player, _exit, ret))
+    return ret
```
Fixed next commit

vgm5 commented on worlds/no100f/__init__.py:
```diff
@@ -0,0 +1,344 @@
+                         ItemNames.MG_Warp]
+
+        # adjust for starting inv prog. items
+        k = 0
+        for item in self.multiworld.precollected_items[self.player]:
+            if item.name in itempool and item.advancement:
+                itempool.remove(item.name)
+                k = k + 1
```
Fixed next commit

vgm5 committed:
Made updates in response to review

Updated various elements of world in response to comments, reformatted some file to align more with PEP 8 Coding standard

vgm5 commented on worlds/no100f/Regions.py:
```diff
@@ -0,0 +1,191 @@
+}
+
+
+def create_regions(world: MultiWorld, options: NO100FOptions, player: int):
+    # create regions
+    world.regions += [
+        create_region(world, player, k, _get_locations_for_region(options, k), v) for k, v in exit_table.items()
+    ]
+
+    # connect regions
+    world.get_entrance(ConnectionNames.start_game, player).connect(world.get_region(RegionNames.hub1, player))
+    for k, v in exit_table.items():
+        if k == RegionNames.menu:
+            continue
+        for _exit in v:
+            exit_regions = _exit.split('->')
+            assert len(exit_regions) == 2
+            target = world.get_region(exit_regions[1], player)
+            world.get_entrance(_exit, player).connect(target)
```
Ultimately I was unable to get this to work smoothly with my implementation, fell back to using original create_regions

vgm5 commented:
Comments addressed, slight reformatting done in some files to more closely align with PEP 8 Coding Style

ScipioWright commented on worlds/no100f/__init__.py:
```diff
@@ -181,8 +183,11 @@ def check_version(cls, opened_zipfile: zipfile.ZipFile) -> bool:
         return True
 
 
+def get_filler_item_name(self) -> str:
+    return self.random.choice("Scooby Snack", "Scooby Snack Box")
```
This is supposed to go in the World class. I usually put it after `create_item` for organization purposes

vgm5 committed:
Relocated get filler item name location

Moved get filler item name to be properly defined in world class

vgm5 committed:
Fixed Minor Logic Bug

S001 Connection changed to check for 2 Progressive Jumps instead of 1 for Snacksanity logic

ScipioWright commented on worlds/no100f/Rules.py:
```diff
@@ -0,0 +1,2777 @@
+
+        # Manor
+        ConnectionNames.i020_i021: lambda player: lambda state: state.has(ItemNames.HelmetPower, player, 1),
+        ConnectionNames.i003_b004: lambda player: lambda state: state.has(ItemNames.SoapPower, player, 1) and state.has(ItemNames.HelmetPower, player, 1),
```
```suggestion
        ConnectionNames.i003_b004: lambda player: lambda state: state.has_all((ItemNames.SoapPower, ItemNames.HelmetPower), player),
```
Probably not necessary to do this refactor, but would recommend using `state.has_all` in-place of two `state.has`.
Also `state.has` defaults the quantity to 1, so if the quantity you're putting in is 1, the quantity can be omitted.

vgm5 replied:
Yea that would definitely make my code a fair bit more readable.

Unfortunately with how large that refactor would be, and how much it depends on a case by case basis I don't think I could justify doing it for now.

ScipioWright commented on worlds/no100f/Rules.py:
```diff
@@ -0,0 +1,2777 @@
+            LocationNames.o006_SN23: lambda player: lambda state: ((state.can_reach(RegionNames.o005, "Region", player) and state.has(ItemNames.HelmetPower, player, 1) and state.has(ItemNames.ProgressiveJump, player, 1)) or state.has(ItemNames.Knight_Key, player, 4) or state.has(ItemNames.Knight_KeyRing, player, 1)),
+            LocationNames.o006_SN24: lambda player: lambda state: ((state.can_reach(RegionNames.o005, "Region", player) and state.has(ItemNames.HelmetPower, player, 1) and state.has(ItemNames.ProgressiveJump, player, 1)) or state.has(ItemNames.Knight_Key, player, 4) or state.has(ItemNames.Knight_KeyRing, player, 1)),
+            LocationNames.o006_SN25: lambda player: lambda state: ((state.can_reach(RegionNames.o005, "Region", player) and state.has(ItemNames.HelmetPower, player, 1) and state.has(ItemNames.ProgressiveJump, player, 1)) or state.has(ItemNames.Knight_Key, player, 4) or state.has(ItemNames.Knight_KeyRing, player, 1)),
+            Locat
```
There's this huge block of basically the same location rule repeatedly.
Would recommend just shoving all of these into a region if they have identical access rules and are all in the same region.
This will make it more performant I think. By how much? Not sure, but probably a fair bit.

vgm5 replied:
To what extent should this be done, that's definitely the biggest chunk of rules out of all of the code - but there are technically loads of places that could be done.

vgm5 replied:
For now I will be committing the code to move those particularly egregious snacks into a separate region.  Will update next commit

ScipioWright replied:
Hopefully it'll make the test gens faster! I usually do 100-player games with all random options to test gen, but then half an hour passed without it finishing placing items, which is a little horrifying imo

vgm5 replied:
I'll the first to admit that my code can be optimized, and that also by design this randomizer is meant to be...extreme.

ScipioWright commented on worlds/no100f/Options.py:
```diff
@@ -0,0 +1,122 @@
+
+
+class SnackCount(Range):
+    """Sets the number of tokens needed if Token Completion Goal is being used"""
```
```suggestion
    """Sets the number of Scooby Snacks needed if Snack Completion Goal is being used"""
```

ScipioWright replied:
Also being able to set it to 1 sounds horrifying

vgm5 replied:
Fixed Next Commit

vgm5 committed:
Made Changes for review comments #2

Moved o006 Snacks into o006_sn subregion to compact and simplify rules in region.

Fix typo in Snack Goal Option

vgm5 committed:
Small Logic Fix

Add restrictions to P005->P004 Connections

ScipioWright reviewed (approved):
Reviewed the code and did test gens.
Did not test webhost.
Did not do a test game.

I don't super agree with the snack goal but eh, it's probably fine.
I think if someone wants a mcguffin hunt goal, they should use the token goal.

vgm5 committed:
Additional Logic Fixes

Various logic fixes to different snacks in the game

vgm5 committed:
Polished Logic

Split up snacks in Clamor 2 and All Scares Upstairs 3 to be less strict with regards to helmet possesion.

Fixed Knight Warp being in logic slightly too early

vgm5 committed:
Minor Snack Bugfix to Client/Logic

Fixed Fear on the Pier logic for 2 slightly hard to grab snacks.

Fixed Snack using wrong Hex ID in Shiver 1

vgm5 committed:
Various Logic Fixes

Added missing logic to a wide range of rooms.

Removed Food improperly randomized as a snack in aghast 2

vgm5 committed:
Massive Rewrite of Location Names

All Snacks on North Side of game have been renamed to be more detailed

Various bugfixes to snack logic and pickups.

vgm5 committed:
Small Logic/Location Fix

Fixed small issue with Location Names causing a dozen location to not exist.

Fixed Cellar 4 logic

vgm5 committed:
Renamed West Side Snacks

Renamed all snacks on west side of game (Hedge, Cliffs, Graveyard)

Fixed Snack Goal being usable without snacksanity.  Fixed Key Pickups in Creepy 2

vgm5 committed:
Finalized Snack Renaming

Renamed Western Map Snacks, Reordered Locations.py to match new total game snack order (Making the Spoiler reasonable to read)

vgm5 committed:
Fixed Client Snack Alignment

Reordered Snack Locations in Client

vgm5 committed:
Added missing snack rules

Added missing snacks for snacks missed in initial dev

vgm5 committed:
Added Missing Snack Rule

One of 4 crate snacks in cellar 1 was missing the rule the other 3 have, fixed

vgm5 committed:
Brough in Patch Changes, bugfixes

From a PR of JacobMIX the patch file can  now be directly opened by Open Patch.

Additionally fixes a logic issue in some cases for All Scares 3.

Furthermore, I have chosen to disable the entry to the Black Knight fight from Cower in the Tower if the player does not have the helmet - this prevents a softlock for that specific check.

vgm5 committed:
Fixed Logic between Cellar and Clamor

Before the logic to go to clamor did not properly check for helmet

vgm5 committed:
Merge remote-tracking branch 'upstream/main'

vgm5 committed:
Fix compatibility with AP 0.6.5

Removed problematic code causing issues with latest AP release

vgm5 committed:
Small bugfixes

Fixed logic issue with backtracking in DaSK, issue where Memory Reads in Cower would lock level up

vgm5 committed:
Small Logic Fixes

Fix a couple logic issues in the Cellar Region

vgm5 committed:
Logic Fixes

Added missing rules for handful of Clamor 1 snacks
